### PR TITLE
Feat(social accounts): Add Tiktok provider

### PR DESCRIPTION
**Pull request by jonasN5**
# Submitting Pull Requests

## General

 - [x] Make sure you use [semantic commit messages](https://seesparkbox.com/foundry/semantic_commit_messages).
       Examples: `"fix(google): Fixed foobar bug"`, `"feat(accounts): Added foobar feature"`.
 - [x] All Python code must formatted using Black, and clean from pep8 and isort issues.
 - [x] If your changes are significant, please update `ChangeLog.rst`.

 ## Provider Specifics

 In case you add a new provider:

- [x] Make sure unit tests are available.
- [x] Add an entry of your provider in `test_settings.py::INSTALLED_APPS` and `docs/installation.rst::INSTALLED_APPS`.
- [x] Add documentation to `docs/providers/<provider name>.rst` and `docs/providers/index.rst` Provider Specifics toctree.
- [x] Add an entry to the list of supported providers over at `docs/overview.rst`. 
-> this file does not exist

Solves https://github.com/pennersr/django-allauth/issues/3703

### CODE REVIEW

File: allauth/socialaccount/providers/oauth2/client.py [hunk 1]
```diff
@@ -11,6 +11,8 @@ class OAuth2Error(Exception):
 
 
 class OAuth2Client(object):
+    CLIENT_ID_PARAMETER = "client_id"
```
Comment (pennersr) [inline allauth/socialaccount/providers/oauth2/client.py:14]:
```suggestion
    client_id_parameter = "client_id"
```

The other constant properties of this class (such as e.g. `scope_delimiter` use just lower case.

File: allauth/socialaccount/providers/tiktok/provider.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    def get_profile_url(self):
+        return self.account.extra_data.get("profile_deep_link")
+
+    def get_avatar_url(self):
+        return self.account.extra_data.get("avatar_url")
+
+    def to_str(self):
+        username = self.get_username()
+        return username or super(TiktokAccount, self).to_str()
```
Comment (pennersr) [inline allauth/socialaccount/providers/tiktok/provider.py:23]:
```suggestion
        return username or super().to_str()
```

File: allauth/socialaccount/providers/tiktok/provider.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+class TiktokProvider(OAuth2Provider):
+    id = "tiktok"
+    name = "Tiktok"
```
Comment (pennersr) [inline allauth/socialaccount/providers/tiktok/provider.py:28]:
```suggestion
    name = "TikTok"
```

File: allauth/socialaccount/providers/tiktok/client.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+from allauth.socialaccount.providers.oauth2.client import OAuth2Client
+
+
+class TiktokOAuth2Client(OAuth2Client):
```
Comment (pennersr) [inline allauth/socialaccount/providers/tiktok/client.py:4]:
```suggestion
class TikTokOAuth2Client(OAuth2Client):
```

File: allauth/socialaccount/providers/tiktok/provider.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+from allauth.socialaccount.providers.base import ProviderAccount
+from allauth.socialaccount.providers.oauth2.provider import OAuth2Provider
+from allauth.socialaccount.providers.tiktok.scopes import TiktokScope
+from allauth.socialaccount.providers.tiktok.views import TiktokOAuth2Adapter
+
+
+class TiktokAccount(ProviderAccount):
```
Comment (pennersr) [inline allauth/socialaccount/providers/tiktok/provider.py:7]:
```suggestion
class TikTokAccount(ProviderAccount):
```

File: allauth/socialaccount/providers/tiktok/provider.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    pkce_enabled_default = False
+
+    def extract_uid(self, data):
+        return str(data["open_id"])
+
+    def extract_common_fields(self, data):
+        return {
+            "username": data.get("username") or data.get("display_name"),
+            "name": data.get("display_name"),
+            "email": None,  # Tiktok does not provide an email
```
Comment (pennersr) [inline allauth/socialaccount/providers/tiktok/provider.py:40]: Just leave the `"email"` key out...

File: allauth/socialaccount/providers/tiktok/scopes.py [hunk 7]
```diff
@@ -0,0 +1,6 @@
+from enum import StrEnum
+
+
+class TiktokScope(StrEnum):
```
Comment (pennersr) [inline allauth/socialaccount/providers/tiktok/scopes.py:4]:
```suggestion
class TikTokScope(StrEnum):
```

**Review by pennersr — CHANGES_REQUESTED** [review]

File: allauth/socialaccount/providers/tiktok/views.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    profile_url = "https://open.tiktokapis.com/v2/user/info/"
+    client_class = TikTokOAuth2Client
+    scope_delimiter = ","
+
+    def get_query_fields(self):
+        fields = ""
+        if TikTokScope.user_info_basic in self.get_provider().get_scope():
+            fields = "open_id,display_name,avatar_url"
+        if TikTokScope.user_info_profile in self.get_provider().get_scope():
+            fields += ",username,profile_deep_link"
```
Comment (pennersr) [inline allauth/socialaccount/providers/tiktok/views.py:25]: If `user_info_basic` is not in scope, but `user_info_profile` is, then `fields` will start using a `","`.

Comment (jonasN5) [reply allauth/socialaccount/providers/tiktok/views.py:25]: I don't think you would make a request without basic scope but fair point

**Comment by pennersr** [comment]
The build is failing due to:

```
 allauth/socialaccount/providers/tiktok/tests.py:1: in <module>
      from typing import override
  E   ImportError: cannot import name 'override' from 'typing' (/opt/hostedtoolcache/Python/3.9.19/x64/lib/python3.9/typing.py)
```

Can you have a look?

**Comment by jonasN5** [comment]
> The build is failing due to:
> 
> ```
>  allauth/socialaccount/providers/tiktok/tests.py:1: in <module>
>       from typing import override
>   E   ImportError: cannot import name 'override' from 'typing' (/opt/hostedtoolcache/Python/3.9.19/x64/lib/python3.9/typing.py)
> ```
> 
> Can you have a look?

Removed, only available as of 3.11

**Comment by pennersr** [comment]
There is also this:

```
      from enum import StrEnum
  E   ImportError: cannot import name 'StrEnum' from 'enum' (/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/enum.py)
```

**Comment by pennersr** [comment]
Just one more thing left (`black`) and we can merge this. Thanks!

**Comment by jonasN5** [comment]
> Just one more thing left (`black`) and we can merge this. Thanks!

For me black doesn't change anything.
Screenshot from 2024-04-24 17-14-19

**Pull request closed without merging**